hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java

```diff
@@ -536,24 +561,76 @@ public boolean scheduleExists(ScheduleID id) {
 
 		List<TokenNftInfo> nftInfos = new ArrayList<>();
 		var uniqueTokensMap = uniqueTokens.get();
+		var uniqueTokenAccountCount = uniqueTokenAccountOwnerships.get().getCount(fromGrpcAccountId(aid));
+
+		var ranges = computeAssociationRanges(0, uniqueTokenAccountCount, start, end);
+
 		uniqueTokenAccountOwnerships.get()
-				.get(fromGrpcAccountId(aid), (int) start, (int) end).forEachRemaining(nftId -> {
-			var nft = uniqueTokensMap.get(nftId);
-			nftInfos.add(
-					TokenNftInfo.newBuilder()
-							.setAccountID(aid)
-							.setCreationTime(nft.getCreationTime().toGrpc())
-							.setNftID(NftID.newBuilder()
-									.setTokenID(nftId.tokenId().toGrpcTokenId())
-									.setSerialNumber(nftId.serialNumber())
-									.build())
-							.setMetadata(ByteString.copyFrom(nft.getMetadata()))
-							.build()
-			);
-		});
+				.get(fromGrpcAccountId(aid), (int) ranges.getKey().longValue(), (int) ranges.getValue().longValue())
+				.forEachRemaining(nftId -> {
+					var nft = uniqueTokensMap.get(nftId);
+					nftInfos.add(buildTokenNft(aid, nft, nftId)
+					);
+				});
+		var tokens = tokenStore.listOfTokensServed(aid);
+		for (var tid : tokens) {
+			var uniqueTokenTreasuryCount = uniqueTokenTreasuryOwnerships.get().getCount(fromGrpcTokenId(tid));
+			ranges = computeAssociationRanges(uniqueTokenAccountCount, uniqueTokenAccountCount + uniqueTokenTreasuryCount, start, end);
+			uniqueTokenTreasuryOwnerships.get()
+					.get(fromGrpcTokenId(tid), (int) ranges.getKey().longValue(), (int) ranges.getValue().longValue())
+					.forEachRemaining(nftId -> {
+						var nft = uniqueTokensMap.get(nftId);
+						nftInfos.add(buildTokenNft(aid, nft, nftId));
+					});
+		}
+
 		return Optional.of(nftInfos);
 	}
 
+	private Pair<Long, Long> computeAssociationRanges(long uniqueOwnerShipCountStart, long uniqueOwnershipCountEnd, long start, long end) {
+		long rangeStart = 0;
+		long rangeEnd = 0;
+		if (uniqueOwnerShipCountStart > 0) {
+			// uniqueTokenTreasury calculations
+			if (uniqueOwnershipCountEnd >= end) {
+				if (uniqueOwnerShipCountStart < end) {
+					if (uniqueOwnerShipCountStart > start) {
+						rangeStart = 0;
+						rangeEnd = end - uniqueOwnerShipCountStart;
+					} else {
+						rangeStart = start - uniqueOwnerShipCountStart;
+						rangeEnd = end - uniqueOwnerShipCountStart;
+					}
```

Comment:
nit: Can we just move `rangeEnd = end - uniqueOwnerShipCountStart` at the bottom of the upper `if` ?